elasticdl/pkg/common_new/tensor.go

```diff
@@ -0,0 +1,177 @@
+	"github.com/tensorflow/tensorflow/tensorflow/go/core/framework/types_go_proto"
+	"reflect"
+	"unsafe"
+)
+
+// Tensor definition
+type Tensor struct {
+	Buffer []byte
+	Dims   []int64
+	Dtype  types_go_proto.DataType
+}
+
+// NewEmptyTensor create an empty n-dim tensor
+func NewEmptyTensor(dim []int64, dtype types_go_proto.DataType) *Tensor {
+	var t = Tensor{
+		Buffer: make([]byte, DimProduct(dim)*int64(DtypeSize[dtype])),
+		Dims:   dim,
+		Dtype:  dtype,
+	}
+	return &t
+}
+
+// NewTensor create a n-dim tensor using exsiting slice
+func NewTensor(slice interface{}, dim []int64) *Tensor {
+	v := reflect.ValueOf(slice)
+	length := v.Len()
+	dtype := SliceTypeToDtype[reflect.TypeOf(slice)]
+	bytelen := length * int(DtypeSize[dtype])
+	if int64(length) != DimProduct(dim) {
+		return nil
+	}
+	sliceHeader := reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(v.Pointer())),
+		Cap:  int(bytelen),
+		Len:  int(bytelen),
+	}
+	var t = Tensor{
+		Buffer: *(*[]byte)(unsafe.Pointer(&sliceHeader)),
+		Dims:   dim,
+		Dtype:  dtype,
+	}
+	return &t
+}
+
+// NewEmptyVector create an empty 1-dim tensor
+func NewEmptyVector(dim int64, dtype types_go_proto.DataType) *Tensor {
+	var t = Tensor{
+		Buffer: make([]byte, dim*int64(DtypeSize[dtype])),
+		Dims:   []int64{dim},
+		Dtype:  dtype,
+	}
+	return &t
+}
+
+// NewVector create an empty 1-dim tensor
+func NewVector(slice interface{}) *Tensor {
+	v := reflect.ValueOf(slice)
+	length := v.Len()
+	dtype := SliceTypeToDtype[reflect.TypeOf(slice)]
+	bytelen := length * int(DtypeSize[dtype])
+	if v.Len() != length {
+		return nil
+	}
+	sliceHeader := reflect.SliceHeader{
+		Data: uintptr(unsafe.Pointer(v.Pointer())),
+		Cap:  int(bytelen),
+		Len:  int(bytelen),
+	}
+	var t = Tensor{
+		Buffer: *(*[]byte)(unsafe.Pointer(&sliceHeader)),
+		Dims:   []int64{int64(length)},
+		Dtype:  dtype,
+	}
+	return &t
+}
+
+// DimProduct get the number of the elements of a tensor of this dim
+func DimProduct(dim []int64) int64 {
+	var size int64 = 1
+	for _, d := range dim {
+		size *= d
+	}
+	return size
+}
+
+// SubTensor get the part reference of the tensor
+func SubTensor(t *Tensor, begin int64, length int64) *Tensor {
+	dsize := int64(DtypeSize[t.Dtype])
+	begin *= dsize
+	var subt = Tensor{
+		Buffer: t.Buffer[begin : begin+length*dsize],
+		Dims:   []int64{length},
+		Dtype:  t.Dtype,
+	}
+	return &subt
+}
+
+// RowOfTensor get the row reference of a 2-dim tensor
+func RowOfTensor(t *Tensor, idx int64) *Tensor {
+	if len(t.Dims) != 2 || idx >= t.Dims[0] {
+		return nil
+	}
+	begin := t.Dims[1] * idx
+	return SubTensor(t, begin, t.Dims[1])
+}
+
+// SetSubTensor set a vector to an index of tensor
+func SetSubTensor(t *Tensor, begin int64, length int64, val *Tensor) {
+	dsize := int64(DtypeSize[t.Dtype])
+	begin *= dsize
+	length *= dsize
+	copy(t.Buffer[begin:begin+length], val.Buffer)
+}
+
+// SetTensorRow set a vector to an index of tensor
+func SetTensorRow(t *Tensor, idx int64, vec *Tensor) {
```

Comment:
These four functions could be taken as the memeber function of struct `Tensor`. How about following definition:

```go
func (t* Tensor) GetSubTensor(begin int64, length int64) *Tensor
func (t *Tensor) SetSubTensor(begin int64, length int64, val *Tensor)

func (t* Tensor) GetRow(idx int64) *Tensor
func (t* Tensor) SetRow(idx int64, vec *Tensor)
```

Reply:
done